app/org/sagebionetworks/bridge/play/controllers/AuthenticationController.java

```diff
@@ -102,7 +87,40 @@ public Result resetPassword() throws Exception {
         authenticationService.resetPassword(passwordReset);
         return okResult("Password has been changed.");
     }
-    
+
+    private Result signInWithRetry(final int retryCounter) throws Exception {
+
+        UserSession session = getSessionIfItExists();
+        if (session != null) {
+            setSessionToken(session.getSessionToken());
+            return okResult(new UserSessionInfo(session));
+        }
+
+        final JsonNode json = requestToJSON(request());
+        final SignIn signIn = parseJson(request(), SignIn.class);
+        final Study study = getStudyOrThrowException(json);
+        try {
+            Logger.info("User " + signIn.getUsername() + " signing in for study " + study.getIdentifier() + ".");
+            session = authenticationService.signIn(study, signIn);
+        } catch(ConsentRequiredException e) {
+            setSessionToken(e.getUserSession().getSessionToken());
+            throw e;
+        } catch(ConcurrentModificationException e) {
+            if (retryCounter > 0) {
+                Logger.info("User " + signIn.getUsername() +
+                        " is having a race condition with signing in for study " + study.getIdentifier() + "." +
+                        " Will retry after 1 second.");
+                // controller.signIn() 95% is < 1000 ms 
+                Thread.sleep(1000);
+                return signInWithRetry(retryCounter - 1);
+            }
+            throw e;
+        }
+
+        setSessionToken(session.getSessionToken());
+        return okResult(new UserSessionInfo(session));
```

Comment:
Optional: I feel like this would be cleaner in the try block, so that every execution path has its own block.